File: Samples/NineSlice/UE4/Assets/NineSlice/NineSliceImage.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;

namespace NineSlice
{
    /// <summary>
    /// Nine slice image control
    /// </summary>
    public class NineSliceImage : ContentControl
    {
        public ImageSource ImageSource
        {
            get { return (ImageSource)GetValue(ImageSourceProperty); }
            set { SetValue(ImageSourceProperty, value); }
        }

        public static readonly DependencyProperty ImageSourceProperty = DependencyProperty.Register(
            "ImageSource", typeof(ImageSource), typeof(NineSliceImage),
            new PropertyMetadata(null, OnSlicesChanged));

        public Thickness Slices
        {
            get { return (Thickness)GetValue(SlicesProperty); }
            set { SetValue(SlicesProperty, value); }
        }

        public static readonly DependencyProperty SlicesProperty = DependencyProperty.Register(
            "Slices", typeof(Thickness), typeof(NineSliceImage),
            new PropertyMetadata(new Thickness(), OnSlicesChanged));

        #region 9-grid slices
        public Rect TopLeftViewbox
        {
            get { return (Rect)GetValue(TopLeftViewboxProperty); }
            set { SetValue(TopLeftViewboxProperty, value); }
        }

        public static readonly DependencyProperty TopLeftViewboxProperty = DependencyProperty.Register(
            "TopLeftViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect TopCenterViewbox
        {
            get { return (Rect)GetValue(TopCenterViewboxProperty); }
            set { SetValue(TopCenterViewboxProperty, value); }
        }

        public static readonly DependencyProperty TopCenterViewboxProperty = DependencyProperty.Register(
            "TopCenterViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect TopRightViewbox
        {
            get { return (Rect)GetValue(TopRightViewboxProperty); }
            set { SetValue(TopRightViewboxProperty, value); }
        }

        public static readonly DependencyProperty TopRightViewboxProperty = DependencyProperty.Register(
            "TopRightViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect CenterLeftViewbox
        {
            get { return (Rect)GetValue(CenterLeftViewboxProperty); }
            set { SetValue(CenterLeftViewboxProperty, value); }
        }

        public static readonly DependencyProperty CenterLeftViewboxProperty = DependencyProperty.Register(
            "CenterLeftViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect CenterViewbox
        {
            get { return (Rect)GetValue(CenterViewboxProperty); }
            set { SetValue(CenterViewboxProperty, value); }
        }

        public static readonly DependencyProperty CenterViewboxProperty = DependencyProperty.Register(
            "CenterViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect CenterRightViewbox
        {
            get { return (Rect)GetValue(CenterRightViewboxProperty); }
            set { SetValue(CenterRightViewboxProperty, value); }
        }

        public static readonly DependencyProperty CenterRightViewboxProperty = DependencyProperty.Register(
            "CenterRightViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect BottomLeftViewbox
        {
            get { return (Rect)GetValue(BottomLeftViewboxProperty); }
            set { SetValue(BottomLeftViewboxProperty, value); }
        }

        public static readonly DependencyProperty BottomLeftViewboxProperty = DependencyProperty.Register(
            "BottomLeftViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect BottomCenterViewbox
        {
            get { return (Rect)GetValue(BottomCenterViewboxProperty); }
            set { SetValue(BottomCenterViewboxProperty, value); }
        }

        public static readonly DependencyProperty BottomCenterViewboxProperty = DependencyProperty.Register(
            "BottomCenterViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));

        public Rect BottomRightViewbox
        {
            get { return (Rect)GetValue(BottomRightViewboxProperty); }
            set { SetValue(BottomRightViewboxProperty, value); }
        }

        public static readonly DependencyProperty BottomRightViewboxProperty = DependencyProperty.Register(
            "BottomRightViewbox", typeof(Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Rect()));
        #endregion

        private static void OnSlicesChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            NineSliceImage image = (NineSliceImage)d;

            double width = 0;
            double height = 0;

            if (image.ImageSource != null)
            {
                width = image.ImageSource.Width;
                height = image.ImageSource.Height;
            }

            double wl = Math.Min(image.Slices.Left, width);
            double wc = Math.Max(width - image.Slices.Left - image.Slices.Right, 0);
            double wr = Math.Min(image.Slices.Right, width);
            double ht = Math.Min(image.Slices.Top, height);
            double hc = Math.Max(height - image.Slices.Top - image.Slices.Bottom, 0);
            double hb = Math.Min(image.Slices.Bottom, height);

            image.TopLeftViewbox = new Rect(0, 0, wl, ht);
            image.TopCenterViewbox = new Rect(wl, 0, wc, ht);
            image.TopRightViewbox = new Rect(wl + wc, 0, wr, ht);

            image.CenterLeftViewbox = new Rect(0, ht, wl, hc);
            image.CenterViewbox = new Rect(wl, ht, wc, hc);
            image.CenterRightViewbox = new Rect(wl + wc, ht, wr, hc);

            image.BottomLeftViewbox = new Rect(0, ht + hc, wl, hb);
            image.BottomCenterViewbox = new Rect(wl, ht + hc, wc, hb);
            image.BottomRightViewbox = new Rect(wl + wc, ht + hc, wr, hb);
        }
    }
}

File: Samples/NineSlice/UE4/Source/NineSlice/NineSliceImage.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "NineSliceImage.h"

#include <NsCore/ReflectionImplement.h>
#include <NsGui/UIElementData.h>
#include <NsGui/PropertyMetadata.h>
#include <NsGui/ImageSource.h>
#include <NsDrawing/Thickness.h>


using namespace NineSlice;


////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::ImageSource* NineSliceImage::GetImageSource() const
{
    return GetValue<Noesis::Ptr<Noesis::ImageSource>>(ImageSourceProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void NineSliceImage::SetImageSource(Noesis::ImageSource* value)
{
    SetValue<Noesis::Ptr<Noesis::ImageSource>>(ImageSourceProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Noesis::Thickness& NineSliceImage::GetSlices() const
{
    return GetValue<Noesis::Thickness>(SlicesProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void NineSliceImage::SetSlices(const Noesis::Thickness& value)
{
    SetValue<Noesis::Thickness>(SlicesProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
static void OnSlicesChanged(Noesis::DependencyObject* d, const Noesis::DependencyPropertyChangedEventArgs&)
{
    NineSliceImage* image = (NineSliceImage*)d;

    float width = 0;
    float height = 0;

    Noesis::ImageSource* imageSource = image->GetImageSource();
    if (imageSource != nullptr)
    {
        width = imageSource->GetWidth();
        height = imageSource->GetHeight();
    }

    const Noesis::Thickness& slices = image->GetSlices();

    float wl = Noesis::Min(slices.left, width);
    float wc = Noesis::Max(width - slices.left - slices.right, 0.0f);
    float wr = Noesis::Min(slices.right, width);
    float ht = Noesis::Min(slices.top, height);
    float hc = Noesis::Max(height - slices.top - slices.bottom, 0.0f);
    float hb = Noesis::Min(slices.bottom, height);

    image->SetValue<Noesis::Rect>(NineSliceImage::TopLeftViewboxProperty, Noesis::Rect(0.0f, 0.0f, wl, ht));
    image->SetValue<Noesis::Rect>(NineSliceImage::TopCenterViewboxProperty, Noesis::Rect(wl, 0.0f, wl + wc, ht));
    image->SetValue<Noesis::Rect>(NineSliceImage::TopRightViewboxProperty, Noesis::Rect(wl + wc, 0.0f, wl + wc + wr, ht));

    image->SetValue<Noesis::Rect>(NineSliceImage::CenterLeftViewboxProperty, Noesis::Rect(0.0f, ht, wl, ht + hc));
    image->SetValue<Noesis::Rect>(NineSliceImage::CenterViewboxProperty, Noesis::Rect(wl, ht, wl + wc, ht + hc));
    image->SetValue<Noesis::Rect>(NineSliceImage::CenterRightViewboxProperty, Noesis::Rect(wl + wc, ht, wl + wc + wr, ht + hc));

    image->SetValue<Noesis::Rect>(NineSliceImage::BottomLeftViewboxProperty, Noesis::Rect(0.0f, ht + hc, wl, ht + hc + hb));
    image->SetValue<Noesis::Rect>(NineSliceImage::BottomCenterViewboxProperty, Noesis::Rect(wl, ht + hc, wl + wc, ht + hc + hb));
    image->SetValue<Noesis::Rect>(NineSliceImage::BottomRightViewboxProperty, Noesis::Rect(wl + wc, ht + hc, wl + wc + wr, ht + hc + hb));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(NineSliceImage, "NineSlice.NineSliceImage")
{
    Noesis::UIElementData* data = NsMeta<Noesis::UIElementData>(Noesis::TypeOf<SelfClass>());

    data->RegisterProperty<Noesis::Ptr<Noesis::ImageSource>>(ImageSourceProperty, "ImageSource",
        Noesis::PropertyMetadata::Create(Noesis::Ptr<Noesis::ImageSource>(), OnSlicesChanged));
    data->RegisterProperty<Noesis::Thickness>(SlicesProperty, "Slices",
        Noesis::PropertyMetadata::Create(Noesis::Thickness(), OnSlicesChanged));

    data->RegisterProperty<Noesis::Rect>(TopLeftViewboxProperty, "TopLeftViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(TopCenterViewboxProperty, "TopCenterViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(TopRightViewboxProperty, "TopRightViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(CenterLeftViewboxProperty, "CenterLeftViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(CenterViewboxProperty, "CenterViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(CenterRightViewboxProperty, "CenterRightViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(BottomLeftViewboxProperty, "BottomLeftViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(BottomCenterViewboxProperty, "BottomCenterViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
    data->RegisterProperty<Noesis::Rect>(BottomRightViewboxProperty, "BottomRightViewbox",
        Noesis::PropertyMetadata::Create(Noesis::Rect()));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Noesis::DependencyProperty* NineSliceImage::ImageSourceProperty;
const Noesis::DependencyProperty* NineSliceImage::SlicesProperty;
const Noesis::DependencyProperty* NineSliceImage::TopLeftViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::TopCenterViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::TopRightViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::CenterLeftViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::CenterViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::CenterRightViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::BottomLeftViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::BottomCenterViewboxProperty;
const Noesis::DependencyProperty* NineSliceImage::BottomRightViewboxProperty;